discoverd2/client/client.go

```diff
@@ -0,0 +1,188 @@
+	return c.c.Delete("/services/" + name)
+}
+
+func (c *Client) Service(name string) Service {
+	return newService(c, name)
+}
+
+func (c *Client) Instances(service string, timeout time.Duration) ([]*Instance, error) {
+	s := c.Service(service)
+	instances, err := s.Instances()
+	if len(instances) > 0 || err != nil && err != c.c.ErrNotFound {
+		return instances, err
+	}
+
+	events := make(chan *Event)
+	stream, err := s.Watch(events)
+	if err != nil {
+		return nil, err
+	}
+	defer stream.Close()
+	// get any current instances
+outer:
+	for event := range events {
+		switch event.Kind {
+		case EventKindCurrent:
+			break outer
+		case EventKindUp:
+			instances = append(instances, event.Instance)
+		}
+	}
+	if len(instances) > 0 {
+		return instances, nil
+	}
+	// wait for an instance to come up
+	for {
+		select {
+		case event, ok := <-events:
+			if !ok {
+				return nil, stream.Err()
+			}
+			if event.Kind != EventKindUp {
+				continue
+			}
+			return []*Instance{event.Instance}, nil
+		case <-time.After(timeout):
+			return nil, ErrTimedOut
+		}
+	}
+}
+
+type service struct {
+	client *Client
+	name   string
+}
+
+func newService(client *Client, name string) Service {
+	return &service{
+		client: client,
+		name:   name,
+	}
+}
+
+func (s *service) Leader() (*Instance, error) {
+	res := &Instance{}
+	return res, s.client.c.Get(fmt.Sprintf("/services/%s/leader", s.name), res)
+}
+
+func (s *service) Instances() ([]*Instance, error) {
+	var res []*Instance
+	return res, s.client.c.Get(fmt.Sprintf("/services/%s/instances", s.name), &res)
+}
+
+func (s *service) Addrs() ([]string, error) {
+	instances, err := s.Instances()
+	if err != nil {
+		return nil, err
+	}
+	addrs := make([]string, len(instances))
+	for i, inst := range instances {
+		addrs[i] = inst.Addr
+	}
+	return addrs, nil
+}
+
+func (s *service) Leaders(leaders chan *Instance) (stream.Stream, error) {
+	res, err := s.client.c.RawReq("GET", fmt.Sprintf("/services/%s/leader", s.name), http.Header{"Accept": {"text/event-stream"}}, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	stream := stream.New()
+	go func() {
+		events := make(chan *Event)
+		eventStream := httpclient.Stream(res, events)
+		defer func() {
+			eventStream.Close()
+			// wait for stream to close to prevent race with Err read
+			for range events {
+			}
+			if err := eventStream.Err(); err != nil {
+				stream.Error = err
+			}
+			close(leaders)
+		}()
+		for {
+			select {
+			case event, ok := <-events:
+				if !ok {
+					return
+				}
+				if event.Kind != EventKindLeader {
+					continue
+				}
+				select {
+				case leaders <- event.Instance:
+				case <-stream.StopCh:
+					return
+				}
+			case <-stream.StopCh:
+				return
+			}
+		}
+	}()
+	return stream, nil
+}
+
+func (s *service) Watch(events chan *Event) (stream.Stream, error) {
+	res, err := s.client.c.RawReq("GET", fmt.Sprintf("/services/%s", s.name), http.Header{"Accept": {"text/event-stream"}}, nil, nil)
```

Comment:
Most of our streaming client code is now really generic. What do you think about a helper method on `httpclient` that would wrap this into

```go
func (s *service) Watch(events chan *Event) (stream.Stream, error) {
    return s.client.c.Stream("GET", fmt.Sprintf("/services/%s", s.name), events)
}
```

Stream itself being something like:
```go
func (c *client) Stream(method, url string, out interface{}) {
    res, err := s.client.c.RawReq("GET", fmt.Sprintf("/services/%s", s.name), http.Header{"Accept": {"text/event-stream"}}, nil, nil)
    if err != nil {
       return nil, err
    }
    return httpclient.Stream(res, out), nil
}
```